### PR TITLE
Document IAggregator and QueryEngine.run a bit

### PR DESCRIPTION
**Pull request by pdobacz**
I was going through `reference` and added some auxiliary comments, which would have made sense for me. Let me know if they're accurate and make sense to have.

### CODE REVIEW

File: src/OpenDiffix.Core/QueryEngine.fs [hunk 1]
```diff
@@ -15,6 +15,32 @@ let rec private extractColumns query =
 
 type QueryResult = { Columns: Column list; Rows: Row list }
 
+// The main entrypoint to the `OpenDiffix.Core` library.
+// Takes in a SQL `statement` as main parameter along with a set of
+// parameters in `queryContext`.
+//
+// Returns the query result as a list of data rows, along with the
+// columns specifications.
```
Comment (edongashi) [inline src/OpenDiffix.Core/QueryEngine.fs:23]: This is quite obvious a few lines above with `type QueryResult = { Columns: Column list; Rows: Row list }`

File: src/OpenDiffix.Core/QueryEngine.fs [hunk 2]
```diff
@@ -15,6 +15,32 @@ let rec private extractColumns query =
...
+//
+// Supports two main modes of operation:
+// I. Standard SQL statement, with `aidColumns` specified within the `queryContext`
+//
+// Will rewrite the statement so that the query result is anonymized according to the
+// `aidColumns` provided. For example, it will append `AND NOT diffix_low_count(aidColumns)`
+// to the `HAVING` SQL clause and rewrite regural SQL aggregators to Diffix aggregators
```
Comment (edongashi) [inline src/OpenDiffix.Core/QueryEngine.fs:30]: regural -> regular

File: src/OpenDiffix.Core/QueryEngine.fs [hunk 3]
```diff
@@ -15,6 +15,32 @@ let rec private extractColumns query =
...
+// (e.g. `count` -> `diffix_low_count`, see below)
+//
+// II. SQL statement with Diffix aggregators, no `aidColumns` specified within the `queryContext`
```
Comment (edongashi) [inline src/OpenDiffix.Core/QueryEngine.fs:33]: Can also be without any custom aggregators i.e. regular query.

File: src/OpenDiffix.Core/QueryEngine.fs [hunk 4]
```diff
@@ -15,6 +15,32 @@ let rec private extractColumns query =
...
+//
+// (**NOTE**) some AID columns may (and should) be used within the Diffix aggregator functions
+//
+// Will not rewrite the statement. The caller is expected to use Diffix aggregator functions
+// manually, in order to anonymize the query result manually.
```
Comment (edongashi) [inline src/OpenDiffix.Core/QueryEngine.fs:38]: Doesn't add too much information

**Comment by edongashi** [comment]
The comments in the runner may as well be added to root `README`, which is outdated anyway and needs some cleaning up. What do you think?

File: src/OpenDiffix.Core/QueryEngine.fs (hunk 1 above)
Comment (pdobacz) [reply src/OpenDiffix.Core/QueryEngine.fs:23]: hm, this is in the vein that "it does that, and doesn't do anything else or return any intermediate result, which just happens to be typed `Row list`, plus `Row` is plain-old-data", which isn't that obvious to anyone doing a first read. I can try to rephrase to emphasis that.

File: src/OpenDiffix.Core/QueryEngine.fs (hunk 2 above)
Comment (pdobacz) [reply src/OpenDiffix.Core/QueryEngine.fs:30]: hm, just noticed vscode spellchecker extension doesn't work for `.fs` files, sorry about that

File: src/OpenDiffix.Core/QueryEngine.fs (hunk 4 above)
Comment (pdobacz) [reply src/OpenDiffix.Core/QueryEngine.fs:38]: which part exactly? The two operation modes where very confusing to me, when I was first confronted with this. This distinction is important to make in my opinion, to capture the somewhat "leaky" abstraction, which is OK to have, but confusing, if not made explicit.

My aim is to make the answer to the question "which of the components is responsible for the anonymization?" hard to miss

**Comment by pdobacz** [comment]
> The comments in the runner may as well be added to root `README`, which is outdated anyway and needs some cleaning up. What do you think?

Good idea, I'll try to draw a line somewhere and coordinate these docs

File: src/OpenDiffix.Core/QueryEngine.fs (hunk 4 above)
Comment (edongashi) [reply src/OpenDiffix.Core/QueryEngine.fs:38]: I see what you mean. We used to call these standard queries and anonymizing queries. OK with you if I tweak the comments a bit to match with our terminology?

File: src/OpenDiffix.Core/QueryEngine.fs (hunk 3 above)
Comment (pdobacz) [reply src/OpenDiffix.Core/QueryEngine.fs:33]: technically true, but then it doesn't make sense to use it, hence not a "main mode of operation". I'll make it clearer that here a "mode of operation that brings the user closer to anonymous results this way or another" is meant.

File: src/OpenDiffix.Core/QueryEngine.fs (hunk 4 above)
Comment (pdobacz) [reply src/OpenDiffix.Core/QueryEngine.fs:38]: hm, calling these `standard` vs `anonymizing` queries doesn't fully make it clear. I want to document the intended use of this function (or the entire lib, since this is its entrypoint) to anyone without relying to terminology coined elsewhere, ELI5 style. But I'm working on a significant revamp, I hope it will read better then

**Comment by pdobacz** [comment]
> > The comments in the runner may as well be added to root `README`, which is outdated anyway and needs some cleaning up. What do you think?
> 
> Good idea, I'll try to draw a line somewhere and coordinate these docs

Actually, I looked at README.md and there isn't any overlapping docs there, so I just pointed the reader to the docs about the entrypoint. I think it's good to keep the library docs at the code level, so that they're easily accessible to anyone who ends up arriving from the `CLI`, `Service` or tests.

I couldn't otherwise find anything off or outdated in the README.md, I think it does what it should do.

**Review by edongashi — APPROVED** [review]
